### PR TITLE
Organise job search results using drop down select options

### PR DESCRIPTION
**Pull request by Renny4Real**
## UX Changes:

---------------------------------------------------------------------------------------------------

### Before(Default View):
Screenshot 2019-09-05 at 16 25 18

### After(Default view):
Screenshot 2019-09-05 at 16 26 31

---------------------------------------------------------------------------------------------------

### Before (Desktop):

Screenshot 2019-09-02 at 17 36 35

### After (Desktop JS enabled):
Screenshot 2019-09-02 at 17 33 32

### After (Desktop JS disabled):
Screenshot 2019-09-02 at 17 33 55

---------------------------------------------------------------------------------------------------

### Before (Mobile):
Screenshot 2019-09-02 at 17 40 03

### After (Mobile JS enabled):
Screenshot 2019-09-02 at 17 32 15

### After (Mobile JS disabled):
Screenshot 2019-09-02 at 17 33 00

---------------------------------------------------------------------------------------------------

**Any suggestions/feedback would be appreciated**

### CODE REVIEW

**Comment by dushan-madetech** [comment]
Would it be possible to expand the sort box on mobile so that it is as wide as the screen i.e. so that there's no whitespace on the right?

File: app/controllers/vacancies_controller.rb [hunk 1]
```diff
@@ -19,13 +19,24 @@ class VacanciesController < ApplicationController
                 :sort_column,
                 :sort_order
 
+  # rubocop:disable Metrics/AbcSize
```
Comment (C-gyorfi) [inline app/controllers/vacancies_controller.rb]: I wonder if this is a good practice 🤔

Comment (dushan-madetech) [reply app/controllers/vacancies_controller.rb]: Rubocop is more of a guideline than a be all end all, so while not a perfect solution I don't have any problems with disabling and enabling specific linting as necessary

File: spec/features/job_seekers_can_sort_vacancies_spec.rb [hunk 2]
```diff
@@ -1,40 +1,91 @@
...
+      expect(page).to have_select(I18n.t('jobs.sort_by'), selected: I18n.t('jobs.oldest_posting'))
+      expect(page.find('.vacancy:eq(1)')).to have_content(fifth_vacancy.job_title)
+      expect(page.find('.vacancy:eq(2)')).to have_content(fourth_vacancy.job_title)
+      expect(page.find('.vacancy:eq(3)')).to have_content(second_vacancy.job_title)
+      expect(page.find('.vacancy:eq(4)')).to have_content(third_vacancy.job_title)
+      expect(page.find('.vacancy:eq(5)')).to have_content(first_vacancy.job_title)
+    end
+
+    scenario 'Can be resorted by soonest expiry date and time when option is selected' do
+      select I18n.t('jobs.latest_posting')
```
[the remark this answers is not in the document]
Comment (Renny4Real) [reply spec/features/job_seekers_can_sort_vacancies_spec.rb]: The default view for the page was to sort by the `soonest closing date`.
This felt unrelated to testing the functionality of the select options.
The added spec was to test the contents of the page could be sorted by the `soonest closing date` using the select option as well.

**Comment by Renny4Real** [comment]
@dushan-madetech That could be a suggestion that we could run past Chris. 
The design guide for this UX change uses a mobile view as its model so there could be a reason for its length.

File: app/assets/javascripts/sortJobList.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+$(document).on("turbolinks:load", function() {
+  $("#jobs_sort").change(function() {
+    this.form.submit();
+  });
+  $("input[data-disable-with=Sort]").hide();
```
[the remark this answers is not in the document]
Comment (Renny4Real) [reply app/assets/javascripts/sortJobList.js]: Changed the selector and tried to pseudocode some unit tests for this.
Couldn't get my head around testing 'DOM manipulation'. (◀️ -- Is that what this is? 🤔)

File: spec/features/job_seekers_can_filter_vacancies_spec.rb [hunk 4]
```diff
@@ -284,7 +284,7 @@
       end
 
       expect(page).to have_content(I18n.t('jobs.filters.clear_filters'))
-      click_on 'Closing date'
+      select I18n.t('jobs.closes_soon')
```
[the remark this answers is not in the document]
Comment (Renny4Real) [reply spec/features/job_seekers_can_filter_vacancies_spec.rb]: I have added the missing line but may I question the reliability of this test?
It still passed after adding the line. I might also not fully understand what this test is describing.

File: config/locales/en.yml [hunk 5]
```diff
@@ -73,6 +74,10 @@ en:
     salary: 'Salary'
     publish_on: 'Date posted'
     expires_on: 'Closing date'
+    closes_soon: 'Least time to apply'
```
[the remark this answers is not in the document]
Comment (Renny4Real) [reply config/locales/en.yml]: Have renamed these.

File: app/controllers/vacancies_controller.rb [hunk 6]
```diff
@@ -19,13 +19,24 @@ class VacanciesController < ApplicationController
...
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
   def index
+    redirect_with_sort(:expires_on, :asc) && return if params[:jobs_sort] == 'closes_soon'
```
[the remark this answers is not in the document]
Comment (Renny4Real) [reply app/controllers/vacancies_controller.rb]: Will update this after giving this an attempt.

File: app/controllers/vacancies_controller.rb (hunk 1 above)
Comment (Renny4Real) [reply app/controllers/vacancies_controller.rb]: I searched for a refactoring solution and ended up following the direction that was already taken in the codebase. i.e There are a number of places where rubocop was disabled.

File: app/controllers/vacancies_controller.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (Renny4Real) [reply app/controllers/vacancies_controller.rb]: Have refactored the places were Rubocop got angry. Thanks @adrianclay

File: spec/features/job_seekers_can_filter_vacancies_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Renny4Real) [reply spec/features/job_seekers_can_filter_vacancies_spec.rb]: This has been clarified. Thanks @adrianclay

File: app/assets/javascripts/sortJobList.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Renny4Real) [reply app/assets/javascripts/sortJobList.js]: Have included a feature spec that would test the javascript instead.

File: app/controllers/vacancies_controller.rb (hunk 1 above)
Comment (Renny4Real) [reply app/controllers/vacancies_controller.rb]: Have now refactored these. 👍

**Comment by dushan-madetech** [comment]
> @dushan-madetech That could be a suggestion that we could run past Chris.
> The design guide for this UX change uses a mobile view as its model so there could be a reason for its length.

If the design guide specified it to look like that on mobile then I have no problems with leaving it as is 😄

File: spec/features/job_seekers_can_sort_vacancies_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Renny4Real) [reply spec/features/job_seekers_can_sort_vacancies_spec.rb]: Update: the default view is now `Newest first`

**Pull request merged**